aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,329 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
   return result;
 }
 
+// helper methods for matrix_exp
+namespace {
+
+// Iterates over first `n_batck_dims` and applies `f` to each element.
+// TODO: this implemtation is not efficient, it simply iteratates over
+// batch elements in a serial for-loop.
+template <typename func_t>
+void batch_apply(at::ArrayRef<Tensor> tensors, int64_t n_batch_dims, const func_t& f) {
+  // no-op if batch is empty
+  if (n_batch_dims <= 0) {
+    return;
+  }
+
+  std::vector<Tensor> tensors_view;
+  for (auto& tensor : tensors) {
+    std::vector<int64_t> tensor_view_sizes;
```

Comment:
Can the sizes vary from tensor to tensor, or do all the tensors come from a larger tensor? In the second case, this could be moved out of the loop.

Reply:
Good point. I will simplify the code.

Reply:
The sizes vary.